### PR TITLE
cmd/snap-confine: skip device cgroup setup when running inside a container

### PR DESCRIPTION
**Pull request by bboozzoo**
Do not attempt to set up a device access filtering through cgroups when we're executing inside a container. Especially, in the case of unprivileged containers, the whole set up may be impossible to execute completely, as running in a separate userns means snap-confine is not even able to mount the bpf filesystem and cannot execute bpf() syscalls.

### CODE REVIEW

**Comment by bboozzoo** [comment]
I've verified this locally, but still need to add a spread test.

File: cmd/libsnap-confine-private/utils.c [hunk 1]
```diff
@@ -261,3 +261,42 @@ bool sc_wait_for_file(const char *path, size_t timeout_sec)
...
+	}
+
+	char container[128] = { 0 };
+
+	if (fgets(container, sizeof(container), in) == NULL) {
+		/* nothing read or other error? */
+		return false;
+	}
+
+	size_t r = strlen(container);
```
Comment (zyga) [inline cmd/libsnap-confine-private/utils.c:283]: I would like to see this as a helper, perhaps `sc_chomp` just to hide the string manipulation bits away from this function.

**Review by zyga — COMMENTED** [review]
This part looks okay.

I was wondering if we want to do something in snapd as well, and especially in `snap debug confinement`

File: cmd/libsnap-confine-private/utils.c [hunk 2]
```diff
@@ -261,3 +261,42 @@ bool sc_wait_for_file(const char *path, size_t timeout_sec)
 	}
 	return false;
 }
+
+const char *run_systemd_container = "/run/systemd/container";
+
+static bool _sc_is_in_container(const char *p)
+{
+	// see what systemd-detect-virt --container does in
+	// https://github.com/systemd/systemd/blob/5dcd6b1d55a1cfe247621d70f0e25d020de6e0ed/src/basic/virt.c#L749-L755
```
Comment (zyga) [inline cmd/libsnap-confine-private/utils.c:270]: Can you reference https://systemd.io/CONTAINER_INTERFACE/ which has a somewhat more stable URL and is higher-level.

File: cmd/libsnap-confine-private/utils.c (hunk 1 above)
Comment (zyga) [reply cmd/libsnap-confine-private/utils.c:283]: @bboozzoo a static function in this file is fine.

**Review by pedronis — APPROVED** [review]
thanks

File: cmd/libsnap-confine-private/utils.c (hunk 1 above)
Comment (bboozzoo) [reply cmd/libsnap-confine-private/utils.c:283]: I'd to do it in a followup. I'll leave a TODO for now.

File: cmd/libsnap-confine-private/utils.c (hunk 2 above)
Comment (bboozzoo) [reply cmd/libsnap-confine-private/utils.c:270]: added

**Review by zyga — APPROVED** [review]
LGTM

**Pull request merged**